libvirt/README.md

```diff
@@ -44,3 +44,222 @@ If you want to set default password for podvm debugging then you can use guestfi
+$ docker build --no-cache -t podvm_libvirt --build-arg BUILDER_IMG=podvm_builder:latest \
+	--build-arg CLOUD_PROVIDER=libvirt -f Dockerfile.podvm .
+```
+
+The qcow2 image file will be created inside the `podvm_libvirt` image and can be extracted as shown below:
+
+```
+$ podvm_container=$(docker create podvm_libvirt:latest /bin/bash)
+$ file_name=$(docker export ${podvm_container} | tar -t | grep podvm)
+$ docker cp ${podvm_container}:${file_name} podvm.qcow2
+$ docker rm ${podvm_container}
+```
+
+Next you will need to create a volume on libvirt's system storage and upload the image content. That volume is used by
+the cloud-api-adaptor program to instantiate a new Pod VM. Still on the `image` directory, you should run the following
+commands:
+
+```
+$ export IMAGE=<full-path-to-qcow2>
+
+$ virsh vol-create-as --pool default --name podvm-base.qcow2 --capacity 20G --allocation 2G --prealloc-metadata --format qcow2
+$ virsh vol-upload --vol podvm-base.qcow2 $IMAGE --pool default --sparse
+```
+
+You should see that the `podvm-base.qcow2` volume was proper created:
+
+```
+$ virsh -c qemu:///system vol-info --pool default podvm-base.qcow2
+Name:           podvm-base.qcow2
+Type:           file
+Capacity:       6.00 GiB
+Allocation:     631.52 MiB
+```
+
+## Install and configure Confidential Containers and cloud-api-adaptor in the cluster
+
+The easiest way to install the cloud-api-adaptor along with Confidential Containers in the cluster is through the
+Kubernetes operator available in the `install` directory of this repository.
+
+Start by creating a public/private RSA key pair that will be used by the cloud-api-provider program, running on the
+cluster workers, to connect with your local libvirtd instance without password authentication. Assume you are in the
+`libvirt` directory, do:
+
+```
+$ cd ../install/overlays/libvirt
+$ ssh-keygen -f ./id_rsa -N ""
+$ cat id_rsa.pub >> ~/.ssh/authorized_keys
+```
+**Note**: ensure that `~/.ssh/authorized_keys` has the right permissions (read/write for the user only) otherwise the
+authentication can silently fail.
+
+You will need to figure out the IP address of your local host (e.g. 192.168.1.107). Then try to remote connect with
+libvirt to check the keys setup is fine, for example:
+
+```
+$ virsh -c "qemu+ssh://$USER@192.168.1.107/system?keyfile=$(pwd)/id_rsa" nodeinfo
+CPU model:           x86_64
+CPU(s):              12
+CPU frequency:       1084 MHz
+CPU socket(s):       1
+Core(s) per socket:  6
+Thread(s) per core:  2
+NUMA cell(s):        1
+Memory size:         32600636 KiB
+```
+
+Now you should finally install the Kubernetes operator in the cluster with the help of the [`install_operator.sh`](./install_operator.sh) script. Ensure that you have your IP address exported in the environment, as shown below, then run the install script:
+
+```
+$ cd ../../../libvirt/
+$ export LIBVIRT_IP="192.168.1.107"
+$ export SSH_KEY_FILE="id_rsa"
+$ ./install_operator.sh
```

Comment:
When I ran this I hit an error:
```
# ./install_operator.sh
kubectl apply -f install/yamls/deploy.yaml
namespace/confidential-containers-system created
serviceaccount/cc-operator-controller-manager created
role.rbac.authorization.k8s.io/cc-operator-leader-election-role created
clusterrole.rbac.authorization.k8s.io/cc-operator-manager-role created
clusterrole.rbac.authorization.k8s.io/cc-operator-metrics-reader created
clusterrole.rbac.authorization.k8s.io/cc-operator-proxy-role created
rolebinding.rbac.authorization.k8s.io/cc-operator-leader-election-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/cc-operator-manager-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/cc-operator-proxy-rolebinding created
configmap/cc-operator-manager-config created
service/cc-operator-controller-manager-metrics-service created
deployment.apps/cc-operator-controller-manager created
customresourcedefinition.apiextensions.k8s.io/ccruntimes.confidentialcontainers.org created
kubectl apply -k install/overlays/libvirt
E0119 08:44:17.502494   56037 memcache.go:255] couldn't get resource list for confidentialcontainers.org/v1beta1: the server could not find the requested resource
configmap/peer-pods-cm-libvirt created
secret/auth-json-secret-libvirt created
secret/peer-pods-secret-libvirt created
secret/ssh-key-secret-libvirt created
daemonset.apps/cloud-api-adaptor-daemonset-libvirt created
error: resource mapping not found for name: "ccruntime-sample-libvirt" namespace: "confidential-containers-system" from "install/overlays/libvirt": no matches for kind "CcRuntime" in version "confidentialcontainers.org/v1beta1"
ensure CRDs are installed first
make: *** [Makefile:88: deploy] Error 1
```
and it finished very quickly, so I don't know if there should have been a wait in there or something? When I checked the pods they had been created:
```
# kubectl get pods -n confidential-containers-system
NAME                                             READY   STATUS    RESTARTS   AGE
cc-operator-controller-manager-fbb5dcf9d-w4jts   2/2     Running   0          59s
cloud-api-adaptor-daemonset-libvirt-j88xv        1/1     Running   0          56s
```
I re-ran it and it worked this time: `ccruntime.confidentialcontainers.org/ccruntime-sample-libvirt created`